rfcs/9999-new-merkle-storage.md

```diff
@@ -0,0 +1,140 @@
+## Summary
+[summary]: #summary
+
+I propose creating a new, generic, system for storing arbitrary Merkleized data.
+It would unify and replace the existing system of mask, database, null, and
+sparse ledgers, as well as the Merkle lists we use.
+
+## Motivation
+
+[motivation]: #motivation
+
+The current system is complex, error prone and hard to debug. It also doesn't
+exploit redundancy if you don't explicitly use masks. For example we expect the
+epoch and snarked ledgers to be very similar, there is probably a power law
+distribution on how active accounts are, the vast majority will not have changed
+at all since the epoch snapshot was taken. But in the current system it'd be
+very difficult to exploit that.
+
+## Detailed design
+
+[detailed-design]: #detailed-design
+
+The system is composed of two layers. At the bottom is the Merkle storage system
+itself. It stores and tracks immutable hashed data, both on disk and in RAM,
+using reference counting to free storage appropriately. There is only one copy
+of any item with a given hash, so sharing is automatic and implicit. The primary
+object in storage is a `Ref` (name pending). A `Ref` represents some hashed
+data which may reference other data by hash, if you have a `Ref`, then you have
+the data with the associated hash. The `Ref` type has two parameters, A) a
+phantom that is either `full` or `partial` indicating whether we are guaranteed
+to have all the referenced data as well, and B) what the type of the data is.
+There is a function that takes a `Ref` and ensures that it and all its
+references are stored on disk. We call this "pinning" the `Ref`.
+
+In the table below, the rows correspond to whether the data is pinned and the
+columns correspond to the value of the "fullness" type parameter. The inner
+elements are what those combinations are in the current system (for ledgers).
+
+|        | Full                             | Partial         |
+|--------|----------------------------------|-----------------|
+| Disk   | `Merkle_ledger.Database`         | N/A             |
+| Memory | `Ledger.Mask.Attached` (to null) | `Sparse_ledger` |
+
+Most (maybe all?) of the in-memory ledgers we use in the daemon are actually
+masks on top of a database ledger, which doesn't fit into the new model
+described in this RFC.
```

Comment:
Would it make sense to distinguish in-memory and on-disk `Ref`s by type, if you still need a ledger atop a database ledger?